File: 0131_PalindromePartitioning.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCode
{
    internal class _0131_PalindromePartitioning
    {
        /*
        
        Given a string s, partition s such that every substring of the partition is a 
        palindrome. Return all possible palindrome partitioning of s.

        Input: s = "aab"
        Output: [["a","a","b"],["aa","b"]]

        Input: s = "a"
        Output: [["a"]]
 
        Constraints:

        1 <= s.length <= 16
        s contains only lowercase English letters.

        */
        public IList<IList<string>> Partition(string s)
        {
            var result = new List<IList<string>>();
            var stack = new List<string>();

            dfs(0);

            return result;

            void dfs(int i)
            {
                if (i >= s.Length)
                {
                    result.Add(stack.ToList());
                    return;
                }

                for (var j = i; j < s.Length; j++)
                {
                    if (IsPalindrome(s, i, j))
                    {
                        stack.Add(s.Substring(i, j - i + 1));
                        dfs(j + 1);
                        stack.RemoveAt(stack.Count - 1);
                    }
                }
            }
        }

        public bool IsPalindrome(string s, int l, int r)
        {
            while (l < r)
            {
                if (s[l] != s[r])
                    return false;
                l++;
                r--;
            }

            return true;
        }
    }
}

File: C++ Solutions/LeetCodeCpp/0131_PalindromePartitioning.cpp
/*

Given a string s, partition s such that every substring of the partition is a
palindrome. Return all possible palindrome partitioning of s.

Input: s = "aab"
Output: [["a","a","b"],["aa","b"]]

Input: s = "a"
Output: [["a"]]

Constraints:

1 <= s.length <= 16
s contains only lowercase English letters.

*/

#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<string> curr;
        vector<vector<string>> result;
        dfs(s, 0, curr, result);
        return result;
    }
private:
    void dfs(string s, int start, vector<string>& curr, vector<vector<string>>& result) {
        if (start == s.size()) {
            result.push_back(curr);
            return;
        }
        for (int i = start; i < s.size(); i++) {
            if (isPalindrome(s, start, i)) {
                string str = s.substr(start, i - start + 1);
                curr.push_back(str);
                dfs(s, i + 1, curr, result);
                curr.pop_back();
            }
        }
    }
    bool isPalindrome(string s, int left, int right) {
        while (left < right) {
            if (s[left] != s[right]) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
};
